# Movie Recommendation System

![](https://i.kym-cdn.com/entries/icons/original/000/026/825/movies-tiles.jpg)

# Recommender System

In our Movie recommender system, we will be implementing Simple, Content and Collaborative filtering. Combining all these models we will be building the final model i.e. Hybrid filtering. Here, we will be using two dataset Full Dataset and Small Dataset.

- Full Dataset:- Is made up of 26,000,000 ratings and 750,000 tag applications submitted by 270,000 users to 45,000 movies. Includes tag genome data across 1,100 tags, with 12 million related ratings.

- Small Dataset:- Comprises of 100,000 reviews and 1,300 tag apps added by 700 users to 9,000 movies.


We will build our Simple Recommender using Full Dataset movies, while the small dataset will be used by all personalized recommendation systems like Collaborative Recommender System, Content Based Recommender and Hybrid Recommender (due to the computing power I possess being very limited). Let us build our simple recommendation system as a first step.

# Simple Recommender

The Simple Recommender provides every user with generalized recommendations based on movie popularity and sometimes genre. The underlying principle behind this recommender is that movies that are more popular and more critically acclaimed would be more likely to be enjoyed by the average viewer.  This model does not offer personalized recommendations based on the user.

This model's implementation is highly trivial. All we need to do is sort our movies based on ratings and popularity and display our list of top films. As an added step, to get the top movies of a specific genre, we can pass into a genre argument.

In [1]:
import pandas as pd
import numpy as np
import warnings; warnings.simplefilter('ignore')

movie_data = pd. read_csv("movies_metadata.csv")
movie_data.head(10)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
5  False                                                NaN  60000000   
6  False                                                NaN  58000000   
7  False                                                NaN         0   
8  False                                                NaN  35000000   
9  False  {'id': 645, 'name': 'James Bond Collection', '...  58000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   
5  [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   
6  [{'id': 35, 'name': 'Comedy'}, {'id': 10749, '...   
7  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
8  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
9  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   

                                       homepage     id    imdb_id  \
0          http://toystory.disney.com/toy-story    862  tt0114709   
1                                           NaN   8844  tt0113497   
2                                           NaN  15602  tt0113228   
3                                           NaN  31357  tt0114885   
4                                           NaN  11862  tt0113041   
5                                           NaN    949  tt0113277   
6                                           NaN  11860  tt0114319   
7                                           NaN  45325  tt0112302   
8                                           NaN   9091  tt0114576   
9  http://www.mgm.com/view/movie/757/Goldeneye/    710  tt0113189   

  original_language               original_title  \
0                en                    Toy Story   
1                en                      Jumanji   
2                en             Grumpier Old Men   
3                en            Waiting to Exhale   
4                en  Father of the Bride Part II   
5                en                         Heat   
6                en                      Sabrina   
7                en                 Tom and Huck   
8                en                 Sudden Death   
9                en                    GoldenEye   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   
5  Obsessive master thief, Neil McCauley leads a ...  ...   1995-12-15   
6  An ugly duckling having undergone a remarkable...  ...   1995-12-15   
7  A mischievous young boy, Tom Sawyer, witnesses...  ...   1995-12-22   
8  International action superstar Jean Claude Van...  ...   1995-12-22   
9  James Bond must unmask the mysterious head of ...  ...   1995-11-16   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': '

In [2]:
movie_data.shape

(45466, 24)

In [3]:
movie_data.describe()

revenue       runtime  vote_average    vote_count
count  4.546000e+04  45203.000000  45460.000000  45460.000000
mean   1.120935e+07     94.128199      5.618207    109.897338
std    6.433225e+07     38.407810      1.924216    491.310374
min    0.000000e+00      0.000000      0.000000      0.000000
25%    0.000000e+00     85.000000      5.000000      3.000000
50%    0.000000e+00     95.000000      6.000000     10.000000
75%    0.000000e+00    107.000000      6.800000     34.000000
max    2.787965e+09   1256.000000     10.000000  14075.000000

In [4]:
movie_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
adult                    45466 non-null object
belongs_to_collection    4494 non-null object
budget                   45466 non-null object
genres                   45466 non-null object
homepage                 7782 non-null object
id                       45466 non-null object
imdb_id                  45449 non-null object
original_language        45455 non-null object
original_title           45466 non-null object
overview                 44512 non-null object
popularity               45461 non-null object
poster_path              45080 non-null object
production_companies     45463 non-null object
production_countries     45463 non-null object
release_date             45379 non-null object
revenue                  45460 non-null float64
runtime                  45203 non-null float64
spoken_languages         45460 non-null object
status                   45379 non-null objec

We will be dropping few features that are not useful for our recommender system

In [5]:
# Dropping unnecessary columns
movie_data.drop(['homepage', 'adult', 'overview', 'poster_path', 'tagline', 'video'], axis=1)

belongs_to_collection    budget  \
0      {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                                    NaN  65000000   
2      {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                                    NaN  16000000   
4      {'id': 96871, 'name': 'Father of the Bride Col...         0   
...                                                  ...       ...   
45461                                                NaN         0   
45462                                                NaN         0   
45463                                                NaN         0   
45464                                                NaN         0   
45465                                                NaN         0   

                                                  genres      id    imdb_id  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     862  tt0114709   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...    8844  tt0113497   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   15602  tt0113228   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   31357  tt0114885   
4                         [{'id': 35, 'name': 'Comedy'}]   11862  tt0113041   
...                                                  ...     ...        ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...  439050  tt6209470   
45462                      [{'id': 18, 'name': 'Drama'}]  111109  tt2028550   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   67758  tt0303758   
45464                                                 []  227506  tt0008536   
45465                                                 []  461257  tt6980792   

      original_language               original_title popularity  \
0                    en                    Toy Story    21.9469   
1                    en                      Jumanji    17.0155   
2                    en             Grumpier Old Men    11.7129   
3                    en            Waiting to Exhale    3.85949   
4                    en  Father of the Bride Part II    8.38752   
...                 ...                          ...        ...   
45461                fa                      رگ خواب   0.072051   
45462                tl          Siglo ng Pagluluwal   0.178241   
45463                en                     Betrayal   0.903007   
45464                en          Satana likuyushchiy   0.003503   
45465                en                     Queerama   0.163015   

                                    production_companies  \
0         [{'name': 'Pixar Animation Studios', 'id': 3}]   
1      [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2      [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3      [{'name': 'Twentieth Century Fox Film Corporat...   
4      [{'name': 'Sandollar Productions', 'id': 5842}...   
...                                                  ...   
45461                                                 []   
45462             [{'name': 'Sine Olivia', 'id': 19653}]   
45463  [{'name': 'American World Pictures', 'id': 6165}]   
45464               [{'name': 'Yermoliev', 'id': 88753}]   
45465                                                 []   

                                    production_countries release_date  \
0      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   
...                                                  ...          ...   
45461             [{'iso_3166_1': 'IR', 'name': 'Iran'}]          NaN   
45462      [{'iso_3166_1': 'PH', 'name': 'Philippines'}]   2011-11-17   
45463  [{'iso_3166_1': 'US', 'name': 'United States o...   200

### Features of our dataset


* **belongs_to_collection:** Name of the franchise the movie belongs to.
* **budget:** The budget of the movie in dollars.
* **genres:** A stringified list of dictionaries that list out all the genres associated with the movie.
* **id:** The ID of the move.
* **imdb_id:** The IMDB ID of the movie.
* **original_language:** The language in which the movie was originally shot in.
* **original_title:** The original title of the movie.
* **popularity:** The Popularity Score assigned by TMDB.
* **production_companies:** A stringified list of production companies involved with the making of the movie.
* **production_countries:** A stringified list of countries where the movie was shot/produced in.
* **release_date:** Theatrical Release Date of the movie.
* **revenue:** The total revenue of the movie in dollars.
* **runtime:** The runtime of the movie in minutes.
* **spoken_languages:** A stringified list of spoken languages in the film.
* **status:** The status of the movie (Released, To Be Released, Announced, etc.)
* **title:** The Official Title of the movie.
* **vote_average:** The average rating of the movie.
* **vote_count:** The number of votes by users, as counted by TMDB.

Genre in our dataset is in form of list of dictionaries.So we need to convert it into list form 

In [6]:
import ast
from ast import literal_eval

#converted list of dictionaries to list
movie_data['genres'] = movie_data['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] 
                                                                       if isinstance(x, list) else [])

We will be using the TMDB Ratings to come up with our **Top Movies Chart.** For that we will use IMDB's *weighted rating* formula to construct top chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = $(\frac{v}{v + m} . R) + (\frac{m}{v + m} . C)$

where,
* *v* is the number of votes for the movie
* *m* is the minimum votes required to be listed in the chart
* *R* is the average rating of the movie
* *C* is the mean vote across the whole report

The next step is to determine an appropriate value for *m*, the minimum votes required to be listed in the chart. We will use **95th percentile** as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

The next step is to decide a suitable value for m, the minimum votes needed to appear in the chart. We are going to be using 95th percentile as our limit. In other words, for a movie to be included in the charts, it must have more votes in the list than at least 95 percent of the movies.

Building our overall Top 250 Chart and we will then define a function to build charts for a particular genre.

In [7]:
vote_count = movie_data[movie_data['vote_count'].notnull()]['vote_count'].astype('int')
vote_average = movie_data[movie_data['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_average.mean()
C

5.244896612406511

In [8]:
m = vote_count.quantile(0.95)
m

434.0

In [9]:
movie_data['year'] = pd.to_datetime(movie_data['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [10]:
top_movies = movie_data[(movie_data['vote_count'] >= m) & (movie_data['vote_count'].notnull()) & (movie_data['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
top_movies['vote_count'] = top_movies['vote_count'].astype('int') # converted float to integer
top_movies['vote_average'] = top_movies['vote_average'].astype('int')
top_movies.shape

(2274, 6)

Therefore a movie has to have at least **434 votes** on TMDB to qualify to be considered for the list. We also see that on a scale of 10, the average rating for a movie on TMDB is **5.244**. **2274** movies are qualified to be on our list.

In [11]:
#applying the IMDB's weighted rating formula 
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [12]:
top_movies['weighted_rating'] = top_movies.apply(weighted_rating, axis=1)

In [13]:
top_movies = top_movies.sort_values('weighted_rating', ascending=False).head(250)

## Top Movies

In [14]:
top_movies.head(15)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1225                                  Back to the Future  1985        6239   
834                                        The Godfather  1972        6024   
1154                             The Empire Strikes Back  1980        5998   
46                                                 Se7en  1995        5915   

       vote_average popularity  \
15480             8    29.1081   
12481             8    123.167   
22879             8    32.2135   
2843              8    63.8696   
4863              8    32.0707   
292               8     140.95   
314               8    51.6454   
7000              8    29.3244   
351               8    48.3072   
5814              8    29.4235   
256               8    42.1497   
1225              8    25.7785   
834               8    41.1093   
1154              8     19.471   
46                8    18.4574   

                                                  genres  weighted_rating  
15480  [Action, Thriller, Science Fiction, Mystery, A...         7.917588  
12481                   [Drama, Action, Crime, Thriller]         7.905871  
22879                [Adventure, Drama, Science Fiction]         7.897107  
2843                                             [Drama]         7.881753  
4863                        [Adventure, Fantasy, Action]         7.871787  
292                                    [Thriller, Crime]         7.868660  
314                                       [Drama, Crime]         7.864000  
7000                        [Adventure, Fantasy, Action]         7.861927  
351                             [Comedy, Drama, Romance]         7.860656  
5814                        [Adventure, Fantasy, Action]         7.851924  
256                 [Adventure, Action, Science Fiction]         7.834205  
1225        [Adventure, Comedy, Science Fiction, Family]         7.820813  
834                                       [Drama, Crime]         7.814847  
1154                [Adventure, Action, Science Fiction]         7.814099  
46                            [Crime, Mystery, Thriller]         7.811669

We observe that at the very top of our list are three Christopher Nolan movies, **Inception**, **The Dark Knight** and **Interstellar**. A strong bias of TMDB users towards specific genres and directors is also seen in the table.

Let us create our feature now, which constructs charts for different genres. 

In [15]:
s = movie_data.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
genre_movie = movie_data.drop('genres', axis=1).join(s)

In [16]:
genre_movie.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
1  False                                                NaN  65000000   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

  popularity  ... runtime                                   spoken_languages  \
0    21.9469  ...    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
0    21.9469  ...    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
0    21.9469  ...    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1    17.0155  ...   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
1    17.0155  ...   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
0  Released                                        NaN  Toy Story  False   
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

  vote_average vote_count  year      genre  
0          7.7     5415.0  1995  Animation  
0          7.7     5415.0  1995     Comedy  
0          7.7     5415.0  1995     Family  
1          6.9     2413.0  1995  Adventure  
1          6.9     2413.0  1995    Fantasy  

[5 rows x 25 columns]

In [17]:
def simple_recommender(genre, percentile=0.95):
    movie_data2 = genre_movie[genre_movie['genre'] == genre]
    vote_count = movie_data2[movie_data2['vote_count'].notnull()]['vote_count'].astype('int')
    vote_average = movie_data2[movie_data2['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_average.mean()
    m = vote_count.quantile(percentile)
    
    top_movies = movie_data2[(movie_data2['vote_count'] >= m) & (movie_data2['vote_count'].notnull()) & (movie_data2['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    top_movies['vote_count'] = movie_data2['vote_count'].astype('int')
    top_movies['vote_average'] = top_movies['vote_average'].astype('int')
    
    top_movies['wr'] = top_movies.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    top_movies = top_movies.sort_values('wr', ascending=False).head(250)
    
    return top_movies

Let's see our system in practice by showing the Top 15 Romance Movies (Romance hardly featured anywhere in our Generic Top Chart despite being one of the most common genres of movies).

## Top Romance Movies

In [18]:
simple_recommender('Romance').head(15)

title  year  vote_count  vote_average popularity  \
351                   Forrest Gump  1994        8147             8    48.3072   
10309  Dilwale Dulhania Le Jayenge  1995         661             9     34.457   
876                        Vertigo  1958        1162             8    18.2082   
40251                   Your Name.  2016        1030             8  34.461252   
883               Some Like It Hot  1959         835             8    11.8451   
1132               Cinema Paradiso  1988         834             8     14.177   
19901                     Paperman  2012         734             8    7.19863   
37863                  Sing Street  2016         669             8  10.672862   
1639                       Titanic  1997        7770             7    26.8891   
19731      Silver Linings Playbook  2012        4840             7    14.4881   
40882                   La La Land  2016        4745             7  19.681686   
23437                   Maleficent  2014        4607             7    19.4674   
22168                          Her  2013        4215             7    13.8295   
20910             The Great Gatsby  2013        3885             7    17.5989   
23512       The Fault in Our Stars  2014        3868             7    16.2747   

             wr  
351    7.869860  
10309  7.582757  
876    7.298862  
40251  7.235471  
883    7.117619  
1132   7.116921  
19901  7.041055  
37863  6.984338  
1639   6.916316  
19731  6.869789  
40882  6.867399  
23437  6.863766  
22168  6.852269  
20910  6.840970  
23512  6.840341

The top romance movie according to our metrics is **Forrest Gump**.Followed by bollywood's movie **Dilwale Dulhania Le Jayenge**.

### Data Pre Processing for Persolinzed Recommender Engines

To build our personalized recommender, we will need to merge our current dataset with the crew and the keyword datasets. Let us prepare this data as our first step.
Also, as mentioned in the introduction, we will be using a subset of all the movies available to us due to limiting computing power available.

In [19]:
credits = pd.read_csv('credits.csv')
keywords = pd.read_csv('keywords.csv')
links = pd.read_csv('links.csv')
links = links[links['tmdbId'].notnull()]['tmdbId'].astype('int')

In [20]:
credits.info()
keywords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
cast    45476 non-null object
crew    45476 non-null object
id      45476 non-null int64
dtypes: int64(1), object(2)
memory usage: 1.0+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46419 entries, 0 to 46418
Data columns (total 2 columns):
id          46419 non-null int64
keywords    46419 non-null object
dtypes: int64(1), object(1)
memory usage: 725.4+ KB


The **credits.csv** file contains the cast and crew information of the movie set, while **keywords.csv** contains the keywords used to describe the movie

In [21]:
movie_data = movie_data.drop([19730, 29503, 35587])
movie_data['id'] = movie_data['id'].astype('int')

small_data = movie_data[movie_data['id'].isin(links)]
small_data.shape

keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')

In [22]:
# mergeing credits and keywords dataset with our small dataset
small_data = small_data.merge(credits, on='id')
small_data = small_data.merge(keywords, on='id')

In [23]:
small_data.shape

(9219, 28)

In [24]:
small_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9219 entries, 0 to 9218
Data columns (total 28 columns):
adult                    9219 non-null object
belongs_to_collection    1688 non-null object
budget                   9219 non-null object
genres                   9219 non-null object
homepage                 2001 non-null object
id                       9219 non-null int32
imdb_id                  9219 non-null object
original_language        9219 non-null object
original_title           9219 non-null object
overview                 9207 non-null object
popularity               9219 non-null object
poster_path              9216 non-null object
production_companies     9219 non-null object
production_countries     9219 non-null object
release_date             9219 non-null object
revenue                  9219 non-null float64
runtime                  9219 non-null float64
spoken_languages         9219 non-null object
status                   9217 non-null object
tagline           

We can observe that cast,crew and keywords columns have been added to our small dataset

In [25]:
small_data

adult                              belongs_to_collection    budget  \
0     False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1     False                                                NaN  65000000   
2     False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3     False                                                NaN  16000000   
4     False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...     ...                                                ...       ...   
9214  False                                                NaN   8000000   
9215  False                                                NaN   1000000   
9216  False                                                NaN  15050000   
9217  False                                                NaN  15000000   
9218  False                                                NaN         0   

                                                 genres  \
0                           [Animation, Comedy, Family]   
1                          [Adventure, Fantasy, Family]   
2                                     [Romance, Comedy]   
3                              [Comedy, Drama, Romance]   
4                                              [Comedy]   
...                                                 ...   
9214                                            [Drama]   
9215                                [Thriller, Romance]   
9216               [Adventure, Drama, History, Romance]   
9217  [Action, Adventure, Drama, Horror, Science Fic...   
9218                               [Documentary, Music]   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
9214                                    NaN  159550  tt0255313   
9215                                    NaN  392572  tt5165344   
9216                                    NaN  402672  tt3859980   
9217                                    NaN  315011  tt4262980   
9218  http://www.thebeatlesliveproject.com/  391698  tt2531318   

     original_language                                     original_title  \
0                   en                                          Toy Story   
1                   en                                            Jumanji   
2                   en                                   Grumpier Old Men   
3                   en                                  Waiting to Exhale   
4                   en                        Father of the Bride Part II   
...                ...                                                ...   
9214                en                     The Last Brickmaker in America   
9215                hi                                             रुस्तम   
9216                hi                                       Mohenjo Daro   
9217                ja                                             シン・ゴジラ   
9218                en  The Beatles: Eight Days a Week - The Touring Y...   

                                               overview  ...    status  \
0     Led by Woody, Andy's toys live happily in his ...  ...  Released   
1     When siblings Judy and Peter discover an encha...  ...  Released   
2     A family wedding reignites the ancient feud be...  ...  Released   
3     Cheated on, mistreated and stepped on, the wom...  ...  Released   
4     Just when George Banks has recovered from his ...  ...  Released   
...                                                 ...  ...       ...   
9214  A man must cope with the loss of his wife and ...  ...  Released   
9215  Rustom Pavri, an honourable officer of the Ind...  ...  Released   


Now we've got our cast, crew, genres and credits in one data frame. Let's wrangle this a bit more with the following axioms: 

1) **Crew** :We're only going to select the director from the team as our element, as the others don't add too much to the movie taste.

2) **Cast**: Choosing Cast is a little trickier. Lesser recognized actors and small roles have no real impact on the perception of a film by people. So we only have to pick the main characters and their respective actors. We will randomly pick the top 3 actors appearing in the credits list.

In [26]:
#Converting columns which are in the form of list of dictionaires to list

small_data['cast'] = small_data['cast'].apply(literal_eval)
small_data['crew'] = small_data['crew'].apply(literal_eval)
small_data['keywords'] = small_data['keywords'].apply(literal_eval)
small_data['cast_size'] = small_data['cast'].apply(lambda x: len(x))
small_data['crew_size'] = small_data['crew'].apply(lambda x: len(x))

In [27]:
#defining a function to get directors from the crew list

def director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [28]:
small_data['director'] = small_data['crew'].apply(director)
small_data['cast'] = small_data['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
small_data['cast'] = small_data['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)
small_data['keywords'] = small_data['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
small_data['cast'] = small_data['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])
small_data['director'] = small_data['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
small_data['director'] = small_data['director'].apply(lambda x: [x,x])

In [29]:
small_data = small_data.reset_index()
titles = small_data['title']
indices = pd.Series(small_data.index, index=small_data['title'])

We must do a little pre-processing of our keywords before we put them to good use. We measure the frequency counts of each keyword appearing in the dataset as a first step.

In [30]:
c = small_data.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
c.name = 'keyword'
c = c.value_counts()
c[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64

Keywords occur in frequencies of between 1 and 610. We have no use for keywords which only occur once. And these can be eliminated . We will finally convert every word to its stem so that words like Dogs and Dog are considered the same.

In [31]:
c = c[c > 1]

In [32]:
from nltk.stem.snowball import SnowballStemmer

stemmer = SnowballStemmer('english') #reduces to root word
stemmer.stem('running')

'run'

In [33]:
def keywords(x):
    words = []
    for i in x:
        if i in c:
            words.append(i)
    return words

In [34]:
small_data['keywords'] = small_data['keywords'].apply(keywords)
small_data['keywords'] = small_data['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
small_data['keywords'] = small_data['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

Now that we are done with data pre processing we can move on to our next recommender engine which is **"Content Based Recommender System"**

# Content Based Recommender System

<img src="https://miro.medium.com/max/1026/1*BME1JjIlBEAI9BV5pOO5Mg.png" alt="Drawing" style="width: 300px;"/>


We suffered from some major limitations in the simple recommendation engine. It offers the same suggestion to everyone, no matter the user's personal taste. If our top 15 list were to be looked at by a person who loves romantic films (and hates action), he/she would probably not like any of the films.

To personalise our recommendations more, We are going to build an engine that takes in a movie that a user currently likes as input. Then it analyzes the contents (storyline, genre, cast, director etc.) of the movie to find out other movies which have similar content. Then it ranks similar movies according to their similarity scores and recommends the most relevant movies to the user.Since we will be using contents to build this engine, this is also known as **Content Based Filtering.**



In [35]:
small_data.columns

Index(['index', 'adult', 'belongs_to_collection', 'budget', 'genres',
       'homepage', 'id', 'imdb_id', 'original_language', 'original_title',
       'overview', 'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count', 'year', 'cast', 'crew', 'keywords',
       'cast_size', 'crew_size', 'director'],
      dtype='object')

In this dataset, we observe that a movie has a lot of additional detail. We don't need them all. So as our feature set (the "Content" of the movie), we choose keywords, cast, genres and director column to use.

To do that we will combined all those features in one column

In [36]:
small_data['combined_features'] = small_data['keywords'] + small_data['cast'] + small_data['director'] + small_data['genres']
small_data['combined_features'] = small_data['combined_features'].apply(lambda x: ' '.join(x))
small_data['combined_features']

0       jealousi toy boy friendship friend rivalri boy...
1       boardgam disappear basedonchildren'sbook newho...
2       fish bestfriend duringcreditssting waltermatth...
3       basedonnovel interracialrelationship singlemot...
4       babi midlifecrisi confid age daughter motherda...
                              ...                        
9214    friendship sidneypoitier wendycrewson jayo.san...
9215    bollywood akshaykumar ileanad'cruz eshagupta t...
9216    bollywood hrithikroshan poojahegde kabirbedi a...
9217    monster godzilla giantmonst destruct kaiju hir...
9218    music documentari paulmccartney ringostarr joh...
Name: combined_features, Length: 9219, dtype: object

Now, we need to represent the combined features as vectors. So we will be using TfidfVectorizer() class from sklearn.feature_extraction.text library to do that.

In [37]:
from sklearn.feature_extraction.text import TfidfVectorizer

#Define a TF-IDF Vectorizer Object. Remove all english stopwords
tfidf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tfidf.fit_transform(small_data['combined_features'])

#### Cosine Similarity

We will be using the Cosine Similarity to calculate a numeric quantity that denotes the similarity between two movies. Mathematically, it is defined as follows:

$cosine(x,y) = \frac{x. y^\intercal}{||x||.||y||} $

Since we have used the TF-IDF Vectorizer, calculating the Dot Product will directly give us the Cosine Similarity Score. Therefore, we will use sklearn's **linear_kernel** instead of cosine_similarities since it is much faster.

In [38]:
from sklearn.metrics.pairwise import linear_kernel
cosine_similarity = linear_kernel(tfidf_matrix, tfidf_matrix)

In [39]:
cosine_similarity

array([[1.        , 0.00545459, 0.00235834, ..., 0.        , 0.        ,
        0.        ],
       [0.00545459, 1.        , 0.        , ..., 0.00568909, 0.00483432,
        0.        ],
       [0.00235834, 0.        , 1.        , ..., 0.00507731, 0.        ,
        0.        ],
       ...,
       [0.        , 0.00568909, 0.00507731, ..., 1.        , 0.02244657,
        0.        ],
       [0.        , 0.00483432, 0.        , ..., 0.02244657, 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

We now have a pairwise cosine similarity matrix for all the movies in our dataset. The next step is to write a function that returns the most similar movies based on the cosine similarity score.

In [40]:
def get_recommendations(title):
    index = indices[title]
    simlarity_scores = list(enumerate(cosine_similarity[index]))
    simlarity_scores = sorted(simlarity_scores, key=lambda x: x[1], reverse=True)
    simlarity_scores = simlarity_scores[1:31]
    movie_indices = [i[0] for i in simlarity_scores]
    return titles.iloc[movie_indices]

In [41]:
get_recommendations('The Dark Knight').head(10)

8031         The Dark Knight Rises
6218                 Batman Begins
6623                  The Prestige
1134                Batman Returns
7659    Batman: Under the Red Hood
2085                     Following
7648                     Inception
4145                      Insomnia
3381                       Memento
1260                Batman & Robin
Name: title, dtype: object

The recommendations seem to have acknowledged other Christopher Nolan movies (due to the director's heavy weighting) and placed them as top recommendations. 

In [42]:
get_recommendations('Pulp Fiction').head(10)

1381            Jackie Brown
8905       The Hateful Eight
5200       Kill Bill: Vol. 2
4903       Kill Bill: Vol. 1
898           Reservoir Dogs
8310        Django Unchained
7280    Inglourious Basterds
6788             Death Proof
616     The Great White Hype
4595                   Basic
Name: title, dtype: object

**Popularity and Ratings**

One thing we see in our recommendation system is that, regardless of ratings and popularity, it recommends movies. It is true that **Batman and Robin** compared to **The Dark Knight** have a lot of similar characters, but it was a bad movie that should not be recommended to anyone.

To achieve that, we will add a function to remove terrible movies and return movies which are successful and have had a strong critical response.

Based on similarity scores, We will take the top 50 movies and measure the vote for the 75th percentile movie. Then using this as the $m$ value, we will use the IMDB formula to calculate the weighted rating of each movie, as we did in the Simple Recommender System.

In [43]:
def content_based_recommendations(title):
    index = indices[title]
    similarity_scores = list(enumerate(cosine_similarity[index]))
    similarity_scores = sorted(similarity_scores, key=lambda x: x[1], reverse=True)
    similarity_scores = similarity_scores[1:51]
    movie_indices = [i[0] for i in similarity_scores]
    
    movies = small_data.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.75)
    top_movies = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    top_movies['vote_count'] = top_movies['vote_count'].astype('int')
    top_movies['vote_average'] = top_movies['vote_average'].astype('int')
    top_movies['weighted_rating'] = top_movies.apply(weighted_rating, axis=1)
    top_movies = top_movies.sort_values('weighted_rating', ascending=False).head(10)
    return top_movies

In [44]:
content_based_recommendations('The Dark Knight')

title  vote_count  vote_average  year  \
7648                   Inception       14075             8  2010   
8613                Interstellar       11187             8  2014   
6623                The Prestige        4510             8  2006   
8871                    Deadpool       11444             7  2016   
8031       The Dark Knight Rises        9263             7  2012   
6218               Batman Begins        7511             7  2005   
8872  Captain America: Civil War        7462             7  2016   
8869                     Ant-Man        6029             7  2015   
7583                    Kick-Ass        4747             7  2010   
7600                  Iron Man 2        6969             6  2010   

      weighted_rating  
7648         7.917588  
8613         7.897107  
6623         7.758148  
8871         6.935872  
8031         6.921448  
6218         6.904127  
8872         6.903532  
8869         6.882142  
7583         6.852979  
7600         5.955732

We can clearly observe that movies recommended to us now have good rating and strong critical response. Also the earlier recommendation of **Batman and Robin** is not present as it has terrible vote_average ,i.e, 4.

In [45]:
content_based_recommendations('Pulp Fiction')

title  vote_count  vote_average  year  \
898                 Reservoir Dogs        3821             8  1992   
8310              Django Unchained       10297             7  2012   
7280          Inglourious Basterds        6598             7  2009   
8846  Kingsman: The Secret Service        6069             7  2015   
4903             Kill Bill: Vol. 1        5091             7  2003   
8905             The Hateful Eight        4405             7  2015   
5200             Kill Bill: Vol. 2        4061             7  2004   
1381                  Jackie Brown        1580             7  1997   
3214                   Unbreakable        1994             6  2000   
6788                   Death Proof        1359             6  2007   

      weighted_rating  
898          7.718986  
8310         6.929017  
7280         6.891679  
8846         6.882867  
4903         6.862133  
8905         6.842588  
5200         6.830542  
1381         6.621790  
3214         5.865027  
6788         5.817225

# Collaborative Filtering Recommender System

<img src="https://miro.medium.com/max/345/1*x8gTiprhLs7zflmEn1UjAQ.png" alt="Drawing" style="width: 400px;"/>

Our content-based recommender suffers from some extreme constraints. It can only recommend movies that are **close** to a certain movie. That is, it is not capable of capturing tastes across genres and making recommendations.

Also, it is does not really gives us personal recommendations as it doesn't take into consideration the users personal taste and biases of that user.

In this recommender system, We will therefore use a technique called **Collaborative Filtering** to make more personalized suggestions to Movie Lovers. Collaborative filtering is based on the concept that it is possible to consider a users similar to me to predict which movies I would like which those users have already watched, but I have not.

We will be implementing two different technique for Collabrative Filtering Recommender Engine:
 
 1) Using Scikit Learn's Surprise Library
 
 2) Using Pearson's Correlation

### Using Scikit Learn's Surprise Library

We will be using the **Surprise** library that uses extremely powerful algorithms like **Singular Value Decomposition (SVD)** to minimise RMSE (Root Mean Square Error) and give great recommendations.

In [46]:
ratings = pd.read_csv("ratings.csv")
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [47]:
ratings.rating.value_counts()


4.0    28750
3.0    20064
5.0    15095
3.5    10538
4.5     7723
2.0     7271
2.5     4449
1.0     3326
1.5     1687
0.5     1101
Name: rating, dtype: int64

From here we can see rating of 4.0 has highest value counts. This means more people rated the movie 4.0.

In [48]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100004 entries, 0 to 100003
Data columns (total 4 columns):
userId       100004 non-null int64
movieId      100004 non-null int64
rating       100004 non-null float64
timestamp    100004 non-null int64
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [49]:
ratings.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

In [50]:
from surprise import Reader, Dataset

reader = Reader()

data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)

In [51]:
from surprise.model_selection import train_test_split

trainset, testset = train_test_split(data, test_size=0.25)

In [52]:
from surprise import SVD, accuracy
svd = SVD()

from surprise.model_selection import cross_validate
cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)


Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8857  0.8991  0.8970  0.9031  0.8948  0.8959  0.0058  
MAE (testset)     0.6800  0.6940  0.6924  0.6952  0.6917  0.6906  0.0055  
Fit time          3.72    3.73    3.74    3.72    3.74    3.73    0.01    
Test time         0.10    0.18    0.10    0.10    0.10    0.12    0.03    


{'test_rmse': array([0.88565157, 0.89911118, 0.8969547 , 0.90314233, 0.89481328]),
 'test_mae': array([0.68001171, 0.69395337, 0.69238824, 0.69520955, 0.69168572]),
 'fit_time': (3.7172257900238037,
  3.731935739517212,
  3.7403717041015625,
  3.7234628200531006,
  3.7383670806884766),
 'test_time': (0.10200214385986328,
  0.1836094856262207,
  0.10000491142272949,
  0.1000216007232666,
  0.10000038146972656)}

For our case, we get a mean Root Mean Sqaure Error of 0.8975 which is more than good enough. Let us now train the dataset and do some predictions.

In [53]:
svd.fit(trainset)

In [54]:
predictions = svd.test(testset)

Let us pick user 30 and check the ratings s/he has given.

In [55]:
ratings[ratings['userId'] == 30]

userId  movieId  rating   timestamp
5048      30        1     4.0   944943070
5049      30        2     2.0   945277634
5050      30        6     4.0   945276746
5051      30        8     4.0   968786809
5052      30       11     4.0   948141296
...      ...      ...     ...         ...
6054      30     6436     5.0  1055788276
6055      30     6440     4.0  1055797723
6056      30     6452     5.0  1055788246
6057      30     6473     4.0  1055797789
6058      30     6565     4.5  1060013871

[1011 rows x 4 columns]

In [56]:
svd.predict(30, 302, 3)

Prediction(uid=30, iid=302, r_ui=3, est=3.8342822025140038, details={'was_impossible': False})

 We get an average prediction of 3.941 for film with ID 302. One surprising aspect of this recommender method is that what the film is (or what it contains) doesn't matter. It operates solely on the basis of an allocated film ID, and attempts to predict ratings based on how the other users predicted the film.

### Collaborative Filtering using Pearson Correlation


Pearson’s Coefficient of correlation is used to get statistical relationship between two variables. In collaborative filtering recommender system, we are going to use Pearson’s correlation to get us the correlation coefficient between similar movies which will help us to get more personalized movie recommendation.

In [57]:
ratings = pd.read_csv('ratings.csv')
movies = pd.read_csv('movies.csv')
ratings = pd.merge(movies,ratings).drop(['genres','timestamp'], axis = 1)
ratings.head()

movieId             title  userId  rating
0        1  Toy Story (1995)       7     3.0
1        1  Toy Story (1995)       9     4.0
2        1  Toy Story (1995)      13     5.0
3        1  Toy Story (1995)      15     2.0
4        1  Toy Story (1995)      19     3.0

In [58]:
user_ratings = ratings.pivot_table(index=['userId'],columns=['title'],values='rating')
user_ratings.head()

title   'Hellboy': The Seeds of Creation (2004)  'Round Midnight (1986)  \
userId                                                                    
1                                           NaN                     NaN   
2                                           NaN                     NaN   
3                                           NaN                     NaN   
4                                           NaN                     NaN   
5                                           NaN                     NaN   

title   'Salem's Lot (2004)  'Til There Was You (1997)  'burbs, The (1989)  \
userId                                                                       
1                       NaN                        NaN                 NaN   
2                       NaN                        NaN                 NaN   
3                       NaN                        NaN                 NaN   
4                       NaN                        NaN                 NaN   
5                       NaN                        NaN                 NaN   

title   'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                      
1                        NaN                          NaN   
2                        NaN                          NaN   
3                        NaN                          NaN   
4                        NaN                          NaN   
5                        NaN                          NaN   

title   *batteries not included (1987)  ...And Justice for All (1979)  \
userId                                                                  
1                                  NaN                            NaN   
2                                  NaN                            NaN   
3                                  NaN                            NaN   
4                                  NaN                            NaN   
5                                  NaN                            NaN   

title   1-900 (06) (1994)  ...  Zoom (2006)  Zootopia (2016)  Zulu (1964)  \
userId                     ...                                              
1                     NaN  ...          NaN              NaN          NaN   
2                     NaN  ...          NaN              NaN          NaN   
3                     NaN  ...          NaN              NaN          NaN   
4                     NaN  ...          NaN              NaN          NaN   
5                     NaN  ...          NaN              NaN          NaN   

title   Zulu (2013)  [REC] (2007)  eXistenZ (1999)  xXx (2002)  \
userId                                                           
1               NaN           NaN              NaN         NaN   
2               NaN           NaN              NaN         NaN   
3               NaN           NaN              NaN         NaN   
4               NaN           NaN              NaN         NaN   
5               NaN           NaN              NaN         NaN   

title   xXx: State of the Union (2005)  ¡Three Amigos! (1986)  \
userId                                                          
1                                  NaN                    NaN   
2                                  NaN                    NaN   
3                                  NaN                    NaN   
4                                  NaN                    NaN   
5                                  NaN                    NaN   

title   À nous la liberté (Freedom for Us) (1931)  
userId                                             
1                                             NaN  
2                                             NaN  
3                                             NaN  
4                                             NaN  
5                                             NaN  

[5 rows x 7062 columns]

In [59]:
#removes movies which has less than 10 users who rated it and fill Nan with 0
user_ratings = user_ratings.dropna(thresh=10,axis=1).fillna(0)

In [60]:
user_ratings

title   'burbs, The (1989)  (500) Days of Summer (2009)  \
userId                                                    
1                      0.0                          0.0   
2                      0.0                          0.0   
3                      0.0                          0.0   
4                      0.0                          0.0   
5                      0.0                          0.0   
...                    ...                          ...   
667                    0.0                          0.0   
668                    0.0                          0.0   
669                    0.0                          0.0   
670                    0.0                          0.0   
671                    0.0                          0.0   

title   ...And Justice for All (1979)  10 Things I Hate About You (1999)  \
userId                                                                     
1                                 0.0                                0.0   
2                                 0.0                                0.0   
3                                 0.0                                0.0   
4                                 0.0                                0.0   
5                                 0.0                                0.0   
...                               ...                                ...   
667                               0.0                                0.0   
668                               0.0                                0.0   
669                               0.0                                0.0   
670                               0.0                                0.0   
671                               0.0                                0.0   

title   101 Dalmatians (1996)  \
userId                          
1                         0.0   
2                         0.0   
3                         0.0   
4                         0.0   
5                         0.0   
...                       ...   
667                       0.0   
668                       0.0   
669                       0.0   
670                       0.0   
671                       0.0   

title   101 Dalmatians (One Hundred and One Dalmatians) (1961)  \
userId                                                           
1                                                     0.0        
2                                                     0.0        
3                                                     0.0        
4                                                     5.0        
5                                                     0.0        
...                                                   ...        
667                                                   0.0        
668                                                   0.0        
669                                                   0.0        
670                                                   0.0        
671                                                   0.0        

title   102 Dalmatians (2000)  12 Angry Men (1957)  12 Years a Slave (2013)  \
userId                                                                        
1                         0.0                  0.0                      0.0   
2                         0.0                  0.0                      0.0   
3                         0.0                  0.0                      0.0   
4                         0.0                  0.0                      0.0   
5                         0.0                  0.0                      0.0   
...                       ...                  ...                      ...   
667                       0.0                  0.0                      0.0   
668                       0.0                  0.0                      0.0   
669                       0.0                  0.0                      0.0   
670                       0.0                  0.0                      0.0   
671   

In [61]:
item_similarity = user_ratings.corr(method='pearson')
item_similarity.head()

title                              'burbs, The (1989)  \
title                                                   
'burbs, The (1989)                           1.000000   
(500) Days of Summer (2009)                  0.049358   
...And Justice for All (1979)                0.183453   
10 Things I Hate About You (1999)            0.082564   
101 Dalmatians (1996)                        0.101948   

title                              (500) Days of Summer (2009)  \
title                                                            
'burbs, The (1989)                                    0.049358   
(500) Days of Summer (2009)                           1.000000   
...And Justice for All (1979)                         0.026848   
10 Things I Hate About You (1999)                     0.230095   
101 Dalmatians (1996)                                 0.078582   

title                              ...And Justice for All (1979)  \
title                                                              
'burbs, The (1989)                                      0.183453   
(500) Days of Summer (2009)                             0.026848   
...And Justice for All (1979)                           1.000000   
10 Things I Hate About You (1999)                       0.040358   
101 Dalmatians (1996)                                   0.141549   

title                              10 Things I Hate About You (1999)  \
title                                                                  
'burbs, The (1989)                                          0.082564   
(500) Days of Summer (2009)                                 0.230095   
...And Justice for All (1979)                               0.040358   
10 Things I Hate About You (1999)                           1.000000   
101 Dalmatians (1996)                                       0.180185   

title                              101 Dalmatians (1996)  \
title                                                      
'burbs, The (1989)                              0.101948   
(500) Days of Summer (2009)                     0.078582   
...And Justice for All (1979)                   0.141549   
10 Things I Hate About You (1999)               0.180185   
101 Dalmatians (1996)                           1.000000   

title                              101 Dalmatians (One Hundred and One Dalmatians) (1961)  \
title                                                                                       
'burbs, The (1989)                                                          0.162694        
(500) Days of Summer (2009)                                                 0.040188        
...And Justice for All (1979)                                               0.334189        
10 Things I Hate About You (1999)                                           0.182573        
101 Dalmatians (1996)                                                       0.353796        

title                              102 Dalmatians (2000)  12 Angry Men (1957)  \
title                                                                           
'burbs, The (1989)                             -0.002466             0.067309   
(500) Days of Summer (2009)                     0.068425             0.156769   
...And Justice for All (1979)                   0.005991             0.283803   
10 Things I Hate About You (1999)               0.163474             0.065632   
101 Dalmatians (1996)                           0.118462             0.045265   

title                              12 Years a Slave (2013)  127 Hours (2010)  \
title                                                                          
'burbs, The (1989)                                0.020832          0.105458   
(500) Days of Summer (2009)                       0.117532          0.380868   
...And Justice for All (1979)                    -0.017748          0.069345   
10 Things I Hate About You (1999)                 0.027600          0.106938   
101 Dalmatians (1996)             

In [62]:
def collaborative_recommender(movie_name,user_rating):
    similarity_score = item_similarity[movie_name]*(user_rating-2.5)
    similarity_score = similarity_score.sort_values(ascending=False)
    
    return similarity_score

In [63]:
collaborative_recommender("17 Again (2009)", 5). head(10)

title
17 Again (2009)                                        2.500000
How to Lose a Guy in 10 Days (2003)                    1.162404
Green Lantern (2011)                                   1.154159
Tangled (2010)                                         1.138901
13 Going on 30 (2004)                                  1.095069
Princess Diaries, The (2001)                           1.020039
27 Dresses (2008)                                      1.008918
Mean Girls (2004)                                      0.918612
Harry Potter and the Deathly Hallows: Part 2 (2011)    0.894475
Holiday, The (2006)                                    0.876778
Name: 17 Again (2009), dtype: float64

In [64]:
my_rating = [("17 Again (2009)", 5), ("101 Dalmatians (1996)", 3), ("(500) Days of Summer (2009)",2)]
similar_movies = pd.DataFrame()

for movie,rating in my_rating:
    similar_movies = similar_movies.append(collaborative_recommender(movie,rating), ignore_index = True)
    
similar_movies
similar_movies.sum().sort_values(ascending=False).head(10)

17 Again (2009)                        2.409609
How to Lose a Guy in 10 Days (2003)    1.090288
13 Going on 30 (2004)                  1.069531
Tangled (2010)                         1.017275
Green Lantern (2011)                   0.989302
Princess Diaries, The (2001)           0.980114
27 Dresses (2008)                      0.970204
101 Dalmatians (1996)                  0.829600
D2: The Mighty Ducks (1994)            0.821359
Holiday, The (2006)                    0.797397
dtype: float64

# Hybrid Recommender System

Hybrid recommender system will brings together the techniques we have implemented in the engines based on the content based and the collaborative filter recommender system. Consideration we need to take to implement hybrid system are as follows: 

- **Input**: User ID and the Movie 

- **Output Title**: Related films sorted by that particular user based on predicted ratings.

In [65]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [66]:
id_map = pd.read_csv('links.csv')[['movieId', 'tmdbId']]
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)
id_map.columns = ['movieId', 'id']
id_map = id_map.merge(small_data[['title', 'id']], on='id').set_index('title')

#Build ID to title mappings
indices_map = id_map.set_index('id')

We will build a hybrid function which will combine the techniques used in content based recommender system and surprise library based Collaborative filtering recommender system.

In [67]:
def hybrid(userId, title):
    
    #Extract the cosine_sim index of the movie
    index = indices[title]
    
    #Extract the TMDB ID of the movie
    tmdbId = id_map.loc[title]['id']
    
    #Extract the movie ID internally assigned by the dataset
    movie_id = id_map.loc[title]['movieId']
    
    #Extract the similarity scores and their corresponding index for every movie from the cosine_sim matrix
    similiarity_scores = list(enumerate(cosine_similarity[int(index)]))
    
    #Sort the index, score in decreasing order of similarity scores
    similiarity_scores = sorted(similiarity_scores, key=lambda x: x[1], reverse=True)
    
    #Select the top 25 tuples, excluding the first 
    similiarity_scores = similiarity_scores[1:26]
    
    #Store the cosine_sim indices of the top 25 movies in a list
    movie_indices = [i[0] for i in similiarity_scores]
    
    #Extract the metadata of the aforementioned movies
    movies = small_data.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year', 'id']]
    
    #Compute the predicted ratings using the SVD filter
    movies['predicted_ratings'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    
    #Sort the movies in decreasing order of predicted rating
    movies = movies.sort_values('predicted_ratings', ascending=False)
    
    #Return the top 10 movies as recommendations
    return movies.head(10)

In [68]:
hybrid(1, 'The Dark Knight')

title  vote_count  vote_average  year  \
3381                                  Memento      4168.0           8.1  2000   
8613                             Interstellar     11187.0           8.1  2014   
7648                                Inception     14075.0           8.1  2010   
6623                             The Prestige      4510.0           8.0  2006   
6218                            Batman Begins      7511.0           7.5  2005   
8001                         Batman: Year One       255.0           7.1  2011   
7583                                 Kick-Ass      4747.0           7.1  2010   
2131                                 Superman      1042.0           6.9  1978   
8334  Batman: The Dark Knight Returns, Part 2       426.0           7.9  2013   
2085                                Following       363.0           7.2  1998   

          id  predicted_ratings  
3381      77           3.783995  
8613  157336           3.543562  
7648   27205           3.488185  
6623    1124           3.398754  
6218     272           3.275979  
8001   69735           3.187929  
7583   23483           3.116289  
2131    1924           3.060907  
8334  142061           3.046984  
2085   11660           3.006152

In [69]:
hybrid(500, 'The Dark Knight')

title  vote_count  vote_average  year  \
3381                                  Memento      4168.0           8.1  2000   
7648                                Inception     14075.0           8.1  2010   
8613                             Interstellar     11187.0           8.1  2014   
7583                                 Kick-Ass      4747.0           7.1  2010   
8031                    The Dark Knight Rises      9263.0           7.6  2012   
8334  Batman: The Dark Knight Returns, Part 2       426.0           7.9  2013   
8001                         Batman: Year One       255.0           7.1  2011   
4145                                 Insomnia      1181.0           6.8  2002   
8478     Justice League: Crisis on Two Earths       152.0           7.1  2010   
6733                                     TMNT       349.0           6.0  2007   

          id  predicted_ratings  
3381      77           3.408939  
7648   27205           3.340383  
8613  157336           3.321429  
7583   23483           3.287319  
8031   49026           3.231979  
8334  142061           3.201520  
8001   69735           3.074207  
4145     320           3.029359  
8478   30061           2.937685  
6733    1273           2.910097

We see that we get different suggestions for different users for our hybrid recommender while the film is the same. Our reviews are also more personalised and tailored to individual users.

## Conclusion:
Thus, We have built four different recommendation system based on different ideas and algorithms. They are as follows:

1)**Simple Recommender**: This system used TMDB Vote Count and Vote Averages overall to create Top Movies Charts, in general and for a particular genre. The IMDB Weighted Ranking System was used to determine ratings for which final sorting was performed.

2)**Content Based Recommender**: We have used contents (storyline, genre, cast, director etc.) of the movie to find out other movies which have similar content.It ranks similar movies according to their similarity scores and recommends the most relevant movies to the user.

3)**Collaborative Filtering Recommender:** We have build two collaborative engines;one that uses the powerful Surprise Library to create a collaborative filter based on a decomposition of a single value. The obtained RMSE was less than 1 and the engine received approximate ratings for a given user and film. And other engine that uses pearson's coefficient algorithm to recommend movies that was liked by user with similar taste.

4)**Hybrid Recommender:** We put together ideas from content and collaborative filtering to create an algorithm that gave recommendations of movies to a specific user based on the average ratings it had internally measured for that user.